src/boosting/mvs.hpp

```diff
@@ -0,0 +1,100 @@
+/*!
+ * Copyright (c) 2017 Microsoft Corporation. All rights reserved.
+ * Licensed under the MIT License. See LICENSE file in the project root for license information.
+ */
+#ifndef LIGHTGBM_BOOSTING_MVS_H_
+#define LIGHTGBM_BOOSTING_MVS_H_
+
+#include <LightGBM/boosting.h>
+#include <LightGBM/utils/array_args.h>
+#include <LightGBM/utils/log.h>
+
+#include <string>
+#include <algorithm>
+#include <chrono>
+#include <cstdio>
+#include <cstdint>
+#include <fstream>
+#include <vector>
+
+#include "gbdt.h"
+#include "score_updater.hpp"
+
+namespace LightGBM {
+
+class MVS : public GBDT {
+ public:
+  MVS();
+
+  ~MVS() {
+  }
+
+  void Init(const Config *config, const Dataset *train_data, const ObjectiveFunction *objective_function,
+            const std::vector<const Metric *> &training_metrics) override {
+    GBDT::Init(config, train_data, objective_function, training_metrics);
+    mvs_lambda_ = config_->mvs_lambda;
+    mvs_adaptive_ = config_->mvs_adaptive;
+    ResetMVS();
+    if (objective_function_ == nullptr) {
+      // use customized objective function
+      size_t total_size = static_cast<size_t>(num_data_) * num_tree_per_iteration_;
+      gradients_.resize(total_size, 0.0f);
+      hessians_.resize(total_size, 0.0f);
+    }
+  }
+
+  void ResetTrainingData(const Dataset *train_data, const ObjectiveFunction *objective_function,
+                         const std::vector<const Metric *> &training_metrics) override {
+    GBDT::ResetTrainingData(train_data, objective_function, training_metrics);
+    ResetMVS();
+  }
+
+  void ResetConfig(const Config *config) override {
+    GBDT::ResetConfig(config);
+    mvs_lambda_ = config_->mvs_lambda;
+    mvs_adaptive_ = config_->mvs_adaptive;
```

Comment:
To keep the logic consistent with ordinary `Bagging`, `need_re_bagging` should be set `true` if the configuration for bagging is changed. So we'd better set `need_re_bagging=true` here if either `mvs_lambda_` or `mvs_adaptive_` is changed. (If `mvs_adaptive_` is kept as `true`, but `mvs_lambda_` is changed, then we can ignore this change, though).